services/iplant-groups/src/iplant_groups/clients/grouper.clj

```diff
@@ -611,25 +642,80 @@
 
 (defn remove-membership-permission
-  [username attribute-def-name-id role-id subject-id action-names]
+  [username attribute-def-name role-name subject-id action-names]
   (assign-remove-permission
     (format-membership-permission-remove-request
-      username attribute-def-name-id role-id subject-id action-names)))
+      username attribute-def-name role-name subject-id action-names)))
+
+(defn- format-all-permission-search-request
+  [username attribute-def-name]
+  {:WsRestGetPermissionAssignmentsRequest
+   (remove-vals nil? {:actAsSubjectLookup (act-as-subject-lookup username)
+                      :wsAttributeDefNameLookups [{:name attribute-def-name}]})})
+
+(defn- get-permission-assign-ids
+  [username attribute-def-name]
+  (->> (format-all-permission-search-request username attribute-def-name)
+       (permission-assignment-search*)
+       (distinct-by :attributeAssignId)
+       (group-by :permissionType)
+       (map-kv (fn [k v] [k (mapv :attributeAssignId v)]))))
+
+(defn- format-permission-assign-id-removal-request
+  [username permission-type ids]
+  {:WsRestAssignPermissionsRequest
+   {:permissionAssignOperation "remove_permission"
+    :actAsSubjectLookup (act-as-subject-lookup username)
+    :permissionType permission-type
+    :wsAttributeAssignLookups (mapv (partial hash-map :uuid) ids)}})
+
+(defn- remove-permission-assign-ids
+  [username permission-type ids]
+  (->> (format-permission-assign-id-removal-request username permission-type ids)
+       (assign-remove-permission)))
+
+(defn- remove-existing-permissions
+  [username attribute-def-name]
+  (->> (get-permission-assign-ids username attribute-def-name)
+       (map (fn [[permission-type ids]] (remove-permission-assign-ids username permission-type ids)))
+       dorun))
+
+(defn- assign-role-permissions
+  [username attribute-def-name new-role-permissions]
+  (let [fmt (fn [[k v]] (format-permission-assign-request v username attribute-def-name true [k]))]
+    (->> (group-by :action_name new-role-permissions)
+         (map-kv (fn [k v] [k (role-permissions (mapv :role_name v))]))
+         (map (comp assign-remove-permission fmt))
+         dorun)))
+
+(defn- assign-membership-permissions
+  [username attribute-def-name new-membership-permissions]
+  (let [fmt (fn [[k v]] (format-permission-assign-request v username attribute-def-name true [k]))]
+    (->> (group-by :action_name new-membership-permissions)
+         (map-kv (fn [k v] [k (membership-permissions (map (juxt :role_name :subject_id) v))]))
+         (map (comp assign-remove-permission fmt))
+         dorun)))
+
+(defn replace-permissions
+  [username attribute-def-name new-role-permissions new-membership-permissions]
+  (remove-existing-permissions username attribute-def-name)
+  (assign-role-permissions username attribute-def-name new-role-permissions)
+  (assign-membership-permissions username attribute-def-name new-membership-permissions))
```

Comment:
I didn't particularly like this implementation of the `PUT /attributes/:attribute-name/permissions` endpoint, but I couldn't find a better way to do it using Grouper web services. We may be able to make a more robust implementation by calling the Grouper API directly, but I'd like to avoid that if we can.